packages/protocol/contracts/stability/BreakerBox.sol

```diff
@@ -0,0 +1,305 @@
+   * @param exchange The address of the exchange to be removed.
+   */
+  function removeExchange(address exchange) external onlyOwner {
+    uint256 exchangeIndex;
+    for (uint256 i = 0; i < exchanges.length; ++i) {
+      if (exchanges[i] == exchange) {
+        exchangeIndex = i;
+        break;
+      }
+    }
+
+    require(exchanges[exchangeIndex] == exchange, "Exchange has not been added");
+
+    uint256 lastIndex = exchanges.length - 1;
+    if (exchangeIndex != lastIndex) {
+      exchanges[exchangeIndex] = exchanges[lastIndex];
+    }
+
+    exchanges.pop();
+
+    delete exchangeTradingModes[exchange];
+    emit ExchangeRemoved(exchange);
+  }
+
+  /**
+   * @notice Sets the trading mode for the specified exchange.
+   * @param exchange The address of the exchange.
+   * @param tradingMode The trading mode that should be set.
+   */
+  function setExchangeTradingMode(address exchange, uint64 tradingMode) public onlyOwner {
+    require(
+      tradingMode == 0 || tradingModeBreaker[tradingMode] != address(0),
+      "Trading mode must be default or have a breaker set"
+    );
+
+    TradingModeInfo memory info = exchangeTradingModes[exchange];
+    require(info.lastUpdatedTime > 0, "Exchange has not been added");
+
+    info.tradingMode = tradingMode;
+    info.lastUpdatedTime = uint64(block.timestamp);
+    info.lastUpdatedBlock = uint128(block.number);
+    exchangeTradingModes[exchange] = info;
+
+    emit TradingModeUpdated(exchange, tradingMode);
+  }
+
+  /* ==================== View Functions ==================== */
+
+  /**
+   * @notice Returns an array of breaker addresses from start to end.
+   * @return An ordered list of breakers.
+   */
+  function getBreakers() external view returns (address[] memory) {
+    return breakers.getKeys();
+  }
+
+  /**
+   * @notice Checks whether a breaker with the specifed address has been added.
+   */
+  function isBreaker(address breaker) public view returns (bool) {
+    return breakers.contains(breaker);
+  }
+
+  /**
+   * @notice Returns addresses of exchanges that have been added.
+   */
+  function getExchanges() external view returns (address[] memory) {
+    return exchanges;
+  }
+
+  /* ==================== Check Breakers ==================== */
+
+  /**
+   * @notice Checks breakers for a specified exchange to determine the trading mode.
+   * @param exchangeAddress The address of the exchange to run the checks for.
+   * @return currentTradingMode Returns an int representing the current trading mode
+   *                            for the specified exchange.
+   */
+  function checkBreakers(address exchangeAddress) external returns (uint256 currentTradingMode) {
+    TradingModeInfo memory info = exchangeTradingModes[exchangeAddress];
+
+    // Last updated should always have a value gt 0
+    // if it doesn't we can assume this exchange has not been added.
+    if (info.lastUpdatedTime == 0) {
+      return 0;
+    }
```

Comment:
An unregistered exchange will always have a last updated time == 0 but an exchange in the default trading mode will always have last updated > 0. 
The intention here is to allow trades from an unregistered exchange. Initially this was a require but it feels like better UX to allow trades instead of reverting. This makes it possible to have the breakerBox deployed and only active for certain exchanges.